src/Pos/Generator/BlockEvent.hs

```diff
@@ -41,13 +91,334 @@ data BlockRollbackResult
+    isBlockEventFailure = \case
+        BlockRollbackFailure -> True
+        _ -> False
+
+data BlockEventRollback' blund = BlockEventRollback
+    { _berInput    :: !(NewestFirst NE blund)
     , _berOutValid :: !BlockRollbackResult
+    } deriving (Functor, Foldable)
+
+makeLenses ''BlockEventRollback'
+
+instance IsBlockEventFailure (BlockEventRollback' blund) where
+    isBlockEventFailure = isBlockEventFailure . view berOutValid
+
+type BlockEventRollback = BlockEventRollback' BlundDefault
+
+data BlockEvent' blund
+    = BlkEvApply (BlockEventApply' blund)
+    | BlkEvRollback (BlockEventRollback' blund)
+    deriving (Functor, Foldable)
+
+instance IsBlockEventFailure (BlockEvent' blund) where
+    isBlockEventFailure = \case
+        BlkEvApply    a -> isBlockEventFailure a
+        BlkEvRollback a -> isBlockEventFailure a
+
+type BlockEvent = BlockEvent' BlundDefault
+
+----------------------------------------------------------------------------
+-- Block event generation
+----------------------------------------------------------------------------
+
+{- |
+  Block indices. They're neither 0-based nor 1-based, they can start from any
+  index (even negative). Therefore it's necessary to adjust them before using
+  for lookup in a data structure.
+-}
+newtype BlockIndex = BlockIndex {getBlockIndex :: Int}
+    deriving (Eq, Ord, Num, Real, Integral, Enum, Read, Show,
+              Buildable, Generic, Typeable, NFData, Hashable, Random)
+
+newtype BlockEventCount = BlockEventCount {getBlockEventCount :: Word64}
+    deriving (Eq, Ord, Num, Real, Integral, Enum, Read, Show,
+              Buildable, Generic, Typeable, NFData, Hashable, Random)
+
+-- | A coefficient in the range [0,1]. Pass it to 'weighted' if you ever get
+-- the chance.
+newtype Chance = Chance {getChance :: Rational}
+
+-- | Generate a boolean that may happen to be of true value.
+byChance :: (Monad m, RandomGen g) => Chance -> RandT g m Bool
+byChance (Chance c) = weighted [(False, 1 - c), (True, c)]
+
+data BlockEventGenParams = BlockEventGenParams
+    { _begpSecrets         :: !AllSecrets
+    , _begpBlockCountMax   :: !BlockCount
+    , _begpBlockEventCount :: !BlockEventCount
+    , _begpRollbackChance  :: !Chance
+    , _begpFailureChance   :: !Chance
     }
 
-makeLenses ''BlockEventRollback
+makeLenses ''BlockEventGenParams
+
+{- |
+  Return the range of block indices as a half-open interval (closed on the
+  lower bound, open on the upper bound). For example, for block indices
+  @[0, -2, 4, 7]@ the range is a tuple @(-2, 8)@. The reasons for the interval
+  to be half-open:
+    * it's possible to encode an empty range as @(k, k)@
+    * to compute the amount of elements, one can use simple subtraction
+-}
+
+getBlockIndexRange ::
+       [BlockEvent' BlockIndex]
+    -> (BlockIndex, BlockIndex)
+getBlockIndexRange =
+    let minMax a = Smg.Option (Just (Smg.Min a, Smg.Max a))
+    in maybe (0, 0) (over _2 (+1)) . coerce . foldMap minMax . Compose
```

Comment:
Each function call here requires some time to think 😉 
Coerce is the one for which it's easy to reduce this time :)

Actually now I think that it took me more than 10 minutes because I was stupid and have never used this stuff before.